lucene/core/src/java/org/apache/lucene/search/ParallelSortedCollector.java

```diff
@@ -0,0 +1,612 @@
+        this.reverseMul = 1;
+        this.comparator = new MultiLeafFieldComparator(comparators, queue.reverseMul);
+      }
+      this.context = context;
+      leafTerminationState = maxScoreTerminator.addLeafState();
+    }
+
+    void countHit() {
+      ++totalHits;
+      // TODO: replace hitsThresholdChecker with something simpler
+      hitsThresholdChecker.incrementHitCount();
+    }
+
+    void collectHitIfCompetitive(int doc) throws IOException {
+      if (reverseMul * comparator.compareBottom(doc) > 0) {
+        comparator.copy(bottom.slot, doc);
+        score = getComparatorValue(bottom.slot);
+        //System.out.printf("leaf=%d doc=%d score=%f\n", context.ord, context.docBase + doc, score);
+        updateBottom(doc);
+        comparator.setBottom(bottom.slot);
+      } else {
+        // The comparator has no score from this document. We can use any noncompetitive value
+        // to induce this leaf to be terminated.
+        //System.out.printf("leaf=%d doc=%d (noncompetitive) score=%f\n", context.ord, context.docBase + doc, score);
+        score = getComparatorValue(bottom.slot) + 1;
+      }
+    }
+
+    void collectAnyHit(int doc, int hitsCollected) throws IOException {
+      // Startup transient: queue hasn't gathered numHits yet
+      int slot = hitsCollected - 1;
+      // Copy hit into queue
+      comparator.copy(slot, doc);
+      // compute the doc's score before it gets moved by updating the priority queue
+      score = getComparatorValue(slot);
+      add(slot, doc);
+      if (queueFull) {
+        comparator.setBottom(bottom.slot);
+      }
+    }
+
+    private double getComparatorValue(int slot) {
+      // We could avoid this cast by genericizing MaxScoreAccumulator and having its Number type
+      // co-vary with this FieldComparator; also - should DocComparator extend
+      // NumericComparator<Integer>?
+      return firstReverseMul * ((Number) firstComparator.value(slot)).doubleValue();
+    }
+
+    void updateTerminationState(int doc) {
+      leafTerminationState.update(score, context.docBase + doc);
+      if ((leafTerminationState.resultCount & maxScoreTerminator.intervalMask) == 0) {
+        //System.out.println("scoreboard update leaf=" + context.ord + " doc=" + context.docBase + "+" + doc + " total=" + totalHits);
+        if (maxScoreTerminator.updateLeafState(leafTerminationState)) {
+          // Stop if across all segments we have collected enough, and our scores are no longer competitive
+          totalHitsRelation = Relation.GREATER_THAN_OR_EQUAL_TO;
+          //System.out.println("scoreboard terminate leaf " + context.ord + " doc=" + context.docBase + "+" + doc + " totalHits=" + totalHits + " score=" + (long) score);
+          throw new CollectionTerminatedException();
+        }
+      }
+    }
+
+    @Override
+    public void setScorer(Scorable scorer) throws IOException {
+      this.scorer = scorer;
+      comparator.setScorer(scorer);
+      minCompetitiveScore = 0f;
```

Comment:
Should we check early termination here too ?
`````
if (maxScoreTerminator.updateLeafState(leafTerminationState)) {
          // Stop if across all segments we have collected enough, and our scores are no longer competitive
          totalHitsRelation = Relation.GREATER_THAN_OR_EQUAL_TO;
          //System.out.println("scoreboard terminate leaf " + context.ord + " doc=" + context.docBase + "+" + doc + " totalHits=" + totalHits + " score=" + (long) score);
          throw new CollectionTerminatedException();
        }
````

Reply:
I don't understand - do you mean check a flag like `canEarlyTerminate` that we have in TopFieldLeafCollector? -- this is only created when early termination is possible (see isApplicable)